File: cpp/test/client_entity_test.cpp
// Generated basic-flow test for the client entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct ClientSetup {
  std::shared_ptr<BluefinShieldconexMgmtSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static ClientSetup client_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/client/ClientTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinShieldconexMgmtSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("client01"), Value("client02"), Value("client03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID", idmap},
    {"BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value("FALSE")},
    {"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value("TRUE");

  ClientSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void client_entity_instance() {
  auto testsdk = BluefinShieldconexMgmtSDK::testSDK();
  auto ent = testsdk->client();
  ASSERT_EQ(ent->getName(), std::string("client"), "entity name");
}

static void client_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"client", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinShieldconexMgmtSDK::testSDK(seed, sdkopts);
  auto se = strsdk->client();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinShieldconexMgmtSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->client();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void client_entity_basic() {
  auto setup = client_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list", "load", "remove"}) {
    auto sk = is_control_skipped("entityOp", std::string("client.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto client_ref01_ent = client->client();
  Value client_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "client"}), "client_ref01"));
  if (!client_ref01_data.is_map()) client_ref01_data = vmap();
  {
    Value client_ref01_data_result = client_ref01_ent->create(Struct::clone(client_ref01_data), Value::undef());
    client_ref01_data = Helpers::toMapAny(client_ref01_data_result);
    if (!client_ref01_data.is_map()) client_ref01_data = vmap();
    ASSERT_TRUE(client_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(client_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value client_ref01_match = vmap();
  Value client_ref01_list = client_ref01_ent->list(Struct::clone(client_ref01_match), Value::undef());
  ASSERT_TRUE(client_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(client_ref01_list), vmap({{"id", getp(client_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

  // LOAD
  Value client_ref01_match_dt0 = vmap({{"id", getp(client_ref01_data, "id")}});
  Value client_ref01_data_dt0_loaded = client_ref01_ent->load(Struct::clone(client_ref01_match_dt0), Value::undef());
  Value client_ref01_data_dt0_load_result = Helpers::toMapAny(client_ref01_data_dt0_loaded);
  ASSERT_TRUE(client_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(client_ref01_data_dt0_load_result, "id"), getp(client_ref01_data, "id"), "expected load result id to match");

  // REMOVE
  {
    Value client_ref01_match_rm0 = vmap({{"id", getp(client_ref01_data, "id")}});
    client_ref01_ent->remove(Struct::clone(client_ref01_match_rm0), Value::undef());
  }

  // LIST
  Value client_ref01_match_rt0 = vmap();
  Value client_ref01_list_rt0 = client_ref01_ent->list(Struct::clone(client_ref01_match_rt0), Value::undef());
  ASSERT_TRUE(client_ref01_list_rt0.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(client_ref01_list_rt0), vmap({{"id", getp(client_ref01_data, "id")}}));
    ASSERT_TRUE(found.empty(), "expected removed entity to not be in list");
  }

}

int main() {
  T_RUN(client_entity_instance);
  T_RUN(client_entity_stream);
  T_RUN(client_entity_basic);
  return sdktest::summary("client_entity_test");
}

File: rust/tests/client_entity_test.rs
// Generated basic-flow test for the client entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use bluefin_shieldconex_mgmt_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use bluefin_shieldconex_mgmt_sdk::utility::voxgigstruct as vs;
use bluefin_shieldconex_mgmt_sdk::{test_sdk, Entity, BluefinShieldconexMgmtEntity, BluefinShieldconexMgmtSDK, Value};

#[test]
fn client_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.client(Value::Noval);
    assert_eq!(ent.get_name(), "client");
}

#[test]
fn client_entity_stream() {
    // stream() runs the list op through the full pipeline and yields each
    // result item. Seed two entities via test mode; with the `streaming`
    // feature active it yields the feature's incremental items, else it
    // falls back to the materialised items — either way every item yields.
    let seed = jo(vec![(
        "entity",
        jo(vec![(
            "client",
            jo(vec![
                ("strm01", jo(vec![("id", Value::str("strm01"))])),
                ("strm02", jo(vec![("id", Value::str("strm02"))])),
            ]),
        )]),
    )]);

    let sdkopts = jo(vec![(
        "feature",
        jo(vec![("streaming", jo(vec![("active", Value::Bool(true))]))]),
    )]);

    let testsdk = test_sdk(seed.clone(), sdkopts);
    let ent = testsdk.client(Value::Noval);
    let items: Vec<Value> = ent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(items.len(), 2, "stream should yield both seeded items");

    // Fallback: streaming inactive still yields both materialised items.
    let plainsdk = test_sdk(seed, Value::Noval);
    let plainent = plainsdk.client(Value::Noval);
    let plain_items: Vec<Value> = plainent
        .stream("list", Value::empty_map(), Value::empty_map())
        .expect("stream failed")
        .collect();
    assert_eq!(plain_items.len(), 2, "fallback stream should yield both items");
}

#[test]
fn client_entity_basic() {
    let setup = client_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["create", "list", "load", "remove"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("client.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();
    // CREATE
    let client_ref01_ent = client.client(Value::Noval);
    let client_ref01_data = to_map(&getp(
        &getpath(&["new", "client"], &setup.data),
        "client_ref01",
    ));

    let client_ref01_data_result = client_ref01_ent
        .create(client_ref01_data.clone(), Value::Noval)
        .expect("create failed");
    let client_ref01_data = to_map(&client_ref01_data_result);
    assert!(
        matches!(client_ref01_data, Value::Map(_)),
        "expected create result to be a map"
    );
    assert!(
        !getp(&client_ref01_data, "id").is_noval(),
        "expected created entity to have an id"
    );

    // LIST
    let client_ref01_match = Value::empty_map();

    let client_ref01_list = client_ref01_ent
        .list(client_ref01_match.clone(), Value::Noval)
        .expect("list failed");
    assert!(
        matches!(client_ref01_list, Value::List(_)),
        "expected list result to be an array"
    );

    let found_item = vs::select(
        &entity_list_to_data(&client_ref01_list),
        &jo(vec![("id", getp(&client_ref01_data, "id"))]),
    );
    assert!(
        !vs::is_empty(&found_item),
        "expected to find created entity in list"
    );

    // LOAD
    let client_ref01_match_dt0 = jo(vec![("id", getp(&client_ref01_data, "id"))]);
    let client_ref01_data_dt0_loaded = client_ref01_ent
        .load(client_ref01_match_dt0.clone(), Value::Noval)
        .expect("load failed");
    let client_ref01_data_dt0_load_result = to_map(&client_ref01_data_dt0_loaded);
    assert!(
        matches!(client_ref01_data_dt0_load_result, Value::Map(_)),
        "expected load result to be a map"
    );
    assert_eq!(
        getp(&client_ref01_data_dt0_load_result, "id"),
        getp(&client_ref01_data, "id"),
        "expected load result id to match"
    );

    // REMOVE
    let client_ref01_match_rm0 = jo(vec![("id", getp(&client_ref01_data, "id"))]);
    client_ref01_ent
        .remove(client_ref01_match_rm0.clone(), Value::Noval)
        .expect("remove failed");

    // LIST
    let client_ref01_match_rt0 = Value::empty_map();

    let client_ref01_list_rt0 = client_ref01_ent
        .list(client_ref01_match_rt0.clone(), Value::Noval)
        .expect("list failed");
    assert!(
        matches!(client_ref01_list_rt0, Value::List(_)),
        "expected list result to be an array"
    );

    let not_found_item = vs::select(
        &entity_list_to_data(&client_ref01_list_rt0),
        &jo(vec![("id", getp(&client_ref01_data, "id"))]),
    );
    assert!(
        vs::is_empty(&not_found_item),
        "expected removed entity to not be in list"
    );

}

fn client_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("client");
    entity_data_file.push("ClientTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("client01"), Value::str("client02"), Value::str("client03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID", idmap.clone()),
        ("BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value::str("FALSE")),
        ("BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value::str("FALSE")),
        ("BLUEFINSHIELDCONEXMGMT_APIKEY", Value::str("NONE")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "BLUEFINSHIELDCONEXMGMT_TEST_CLIENT_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    let live = getp(&env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            &ja(vec![jo(vec![("apikey", getp(&env, "BLUEFINSHIELDCONEXMGMT_APIKEY"))]), extra]),
            None,
        );
        BluefinShieldconexMgmtSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}
